api/application.go

```diff
@@ -678,33 +669,35 @@ func (h ApplicationHandler) FactCreate(ctx *gin.Context) {
 // @success 204
 // @router /applications/{id}/facts/{key} [put]
 // @param id path string true "Application ID"
-// @param key path string true "Fact key"
-// @param source path string true "Fact source"
-// @param fact body api.Fact true "Fact data"
+// @param key path string true "Qualified fact"
+// @param fact body object true "Fact value"
 func (h ApplicationHandler) FactPut(ctx *gin.Context) {
 	id := h.pk(ctx)
-	r := &Fact{}
-	err := h.Bind(ctx, &r)
-	if err != nil {
-		_ = ctx.Error(err)
-		return
-	}
 	app := &model.Application{}
 	result := h.DB(ctx).First(app, id)
 	if result.Error != nil {
 		_ = ctx.Error(result.Error)
 		return
 	}
 
-	key := ctx.Param(Key)
-	source := ctx.Param(Source)[1:]
-	value, _ := json.Marshal(r.Value)
+	source, key := qualifiedFact(ctx.Param(Key))
+	if key == "" {
+		h.FactReplace(ctx, source)
+		return
+	}
+	f := Fact{}
+	err := h.Bind(ctx, &f.Value)
+	if err != nil {
+		_ = ctx.Error(err)
+		return
+	}
+
 	result = h.DB(ctx).First(&model.Fact{}, "ApplicationID = ? AND Key = ? AND source = ?", id, key, source)
 	if result.Error == nil {
 		result = h.DB(ctx).
 			Model(&model.Fact{}).
 			Where("ApplicationID = ? AND Key = ? AND source = ?", id, key, source).
-			Update("Value", value)
+			Update("Value", f.Value)
```

Comment:
Since the model is being fetched (and it's super small), seems like it would be simpler to:
```
m := &model.Fact{}
db.First(m)
m.Value = f.Value
db.Save(m)
```